chore(maintenance): add powertools to user-agent in SDK clients

am29d opened the pull request:
<!--- 
Instructions:
1. Make sure you follow our Contributing Guidelines: https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/CONTRIBUTING.md
2. Please follow the template, and do not remove any section/item. If something is not applicable leave it empty, but leave it in the PR. 
3. -->

## Description of your changes

In this PR we introduce our custom middleware that can be added to any middleware stack of AWS SDK v3 clients. We decided to use middleware instead of `customUserAgent` options, because we might get SDK client from the user. Because `config` field of the client is readonly, we can't add change it, after the client has been initialised. 

Any feature using an SDK client can call like this:

```ts
const ddbClient = new DynamoDBClient(options);

addUserAgentMiddleware(ddbClient, 'my-feature);
```

this will modify the `user-agent` field by appending  `PT/my-feature/1.10.0 PTEnv/NA`  at the end of the value. 

I was about to add the new functionality to idempotency, but we will do it in a separate PR. Meanwhile, I have removed the singleton for the client in the `DynamoDBPersistenceLayer.ts`, similar how we use it parameters utility. 

<!---
Include here a summary of the change.

Please include also relevant motivation and context.

Add any applicable code snippets, links, screenshots, or other resources
that can help us verify your changes.
-->

### Related issues, RFCs

<!--- 
Add here the number (i.e. #42) to the Github Issue or RFC that is related to this PR.

Don't include any other text, otherwise the Github Issue will not be detected.

Note: If no issue is present the PR might get blocked and not be reviewed.
-->
**Issue number:** closes #1533

## Checklist

- [x] [My changes meet the tenets criteria](https://docs.powertools.aws.dev/lambda-typescript/#tenets)
- [x] I have performed a *self-review* of my own code
- [x] I have *commented* my code where necessary, particularly in areas that should be flagged with a TODO, or hard-to-understand areas
- [ ] I have made corresponding changes to the *documentation*
- [x] My changes generate *no new warnings*
- [x] I have *added tests* that prove my change is effective and works
- [x] The PR title follows the [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/.github/semantic.yml#L2)

### Breaking change checklist

***Is it a breaking change?:*** NO

- [ ] I have documented the migration process
- [ ] I have added, implemented necessary warnings (if it can live side by side)

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

dreamorosi commented on packages/idempotency/src/persistence/DynamoDBPersistenceLayer.ts:
```diff
@@ -230,14 +223,6 @@ class DynamoDBPersistenceLayer extends BasePersistenceLayer {
 
-  private getClient(): DynamoDBClient {
-    if (!this.client) {
-      this.client = new DynamoDBClient(this.clientConfig);
```
The idea here was to initialize the client lazily.

Couldn't we have added the middleware at this point?

```ts
if (!this.client) {
  this.client = new DynamoDBClient(this.clientConfig);
  this.client.middleware.add(...);
}
```

and do the same in the `constructor` with the one that they pass us?

am29d replied:
I wanted to have one place in constructor to do that. Is the performance noticeable?

dreamorosi replied:
I think it's not, let's go with your version.

Let's streamline the constructor altogether. Basically let's do the same as the Parameters util & remove the `private client!` assertion, since we know it'll always be set before the constructor returns.

am29d replied:
Alright, I'll revert to the previous version with lazy init.

dreamorosi commented on packages/commons/src/version.ts:
```diff
@@ -0,0 +1,2 @@
+// this file is auto generated, do not modify
+module.exports = { version: '1.10.0' };
```
```suggestion
export const PT_VERSION = '1.10.0';
```

dreamorosi replied:
This way we are not mixing exports type, the one with `module.exports` is CJS only.

Also if you export it with the final name, you don't have to alias it with `as PT_VERSION` when you import it, same for the object vs `string`

am29d replied:
agree

dreamorosi commented on packages/commons/src/middleware/userAgentMiddleware.ts:
```diff
@@ -0,0 +1,38 @@
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-nocheck
```
I don't feel very comfortable with this.

The easy way would be to just import `@aws-sdk/types` which has all the types we need. However I'm not sure we should be adding it to the bundle.

After the release I'll give it a try and see if I can do some simplified types to avoid the `@ts-nocheck` to the entire file

am29d replied:
I did not manage to add something like `T extends Client`, thus the shortcut. Curious to see how it can be fixed and fully agree that it's not the best way to handle it.

dreamorosi replied:
I think worst case scenario we can add `// @ts-ignore` to specific lines if needed, but keep the rest of the file typed. 

Let's see, it might not be worth creating custom types to avoid issues in case the SDK changes the interfaces. What do you think?

am29d replied:
I have added explicit `// @ts-ignore` for two lines instead of entire file.

Do you mean to create a union type with all the clients we currently support? Something like:

`type SupportedSdkClients = LambdaClient | DynamoDBClient | S3Client;` 

It would imply that we use the `@aws-sdk/types` package, right? Or is there other way to narrow the types of the clients passed to the middleware?

dreamorosi commented on packages/commons/src/index.ts:
```diff
@@ -5,3 +5,4 @@ export * as ContextExamples from './samples/resources/contexts';
 export * as Events from './samples/resources/events';
 export * from './types/middy';
 export * from './types/utils';
+export * from './middleware';
```
This was excluded purposefully. 

The `middleware` folder is specific to Middy-related stuff, which might/might not bring some dependencies.

We don't want to put this in the default export and we want consumers to be explicitly importing the path `@aws-lambda-powertools/commons/lib/middleware` (for now, `@aws-lambda-powertools/commons/middleware` in v2).

dreamorosi replied:
They are both middleware, but of two different kinds, I would consider creating another subfolder called `user-agent` or something that points to the AWS SDK instead.

am29d replied:
Good point, I will put it on the top level

dreamorosi commented on packages/commons/tests/unit/userAgentMiddleware.test.ts:
```diff
@@ -0,0 +1,56 @@
+describe('Function: addUserAgentMiddleware', () => {
+  it('adds powertools user agent to request header at the end', async () => {
+    const lambdaClient = new LambdaClient({
+      logger: console,
```
I think this line is not needed anymore

am29d commented:
* Fixed tests
* added try/catch so we don't block user code if adding user agent fails
* added parametrised tests for each client we use in our utilities

dreamorosi commented on packages/commons/tests/unit/userAgentMiddleware.test.ts:
```diff
@@ -0,0 +1,109 @@
+  },
+};
+
+describe('Given a client of instance: ', () => {
```
Very clean, I like it

dreamorosi reviewed (approved):
Thanks for working on this.

I have run the integration tests on this branch and [they are passing](https://github.com/aws-powertools/powertools-lambda-typescript/actions/runs/5447643933).
